File: src/domain/trajectory-monitor.js
// ═══════════════════════════════════════════════════════════
// TrajectoryMonitor — desvio em metros vs volta de referência
// ═══════════════════════════════════════════════════════════
// Fecha o loop de avaliação das lições L001 (Referência Fixa) e
// L007 (Curva Cega) do P1 Coach. Recebe a sequência de samples
// de UM trecho (curva) + a sequência da MESMA curva na volta de
// referência, e devolve um relatório com:
//
//   • brakingPointDeviationM  — desvio do ponto de freio (m)
//   • turnInPointDeviationM   — desvio do ponto de giro (m)
//   • throttlePointDeviationM — desvio do ponto de tração (m)
//   • entryDeviationM         — desvio da posição na entrada (m)
//   • apexDeviationM          — desvio da posição no apex (m)
//
// Detecção:
//   • freio   = primeira amostra com accLong < BRAKE_G_THRESHOLD
//               OU (sem IMU) primeira queda sustentada de kmh
//   • giro    = primeira amostra com |gyroAlpha| > YAW_THRESHOLD
//               OU mudança de course além de COURSE_DELTA
//   • tração  = primeira amostra após apex com accLong > THROTTLE_G_THRESHOLD
//
// Distância em metros: prefere `track.geoAncoras` (cálculo geográfico
// puro). Se ausentes, usa fallback via lat/lng → metros (haversine
// aproximado para áreas pequenas <5km — autódromo).
//
// Módulo é PURO: não toca DB, não toca DOM, não usa Dexie. Recebe
// tudo por parâmetro. O caller (cockpit/UI) carrega refSamples.

import { Confidence } from './lesson-schema.js';

// ── Limiares (mesmas constantes de fase-curva.js, mantidos em sincronia) ──
const BRAKE_G_THRESHOLD     = -0.20;   // accLong abaixo disso = frenagem
const THROTTLE_G_THRESHOLD  = +0.15;   // accLong acima disso = aceleração
const YAW_THRESHOLD_DEG_S   = 15;      // gyroAlpha acima disso = giro
const COURSE_DELTA_DEG      = 6;       // delta de course entre samples consecutivos
const KMH_DROP_PER_SEC      = 8;       // fallback sem IMU: queda de kmh/s
const MIN_SAMPLES           = 4;       // abaixo disso, monitor desiste
const M_PER_DEG_LAT         = 111000;

// ─── Conversão geográfica ───────────────────────────────────

/**
 * Distância em metros entre dois pontos lat/lng (aproximação plana,
 * válida para distâncias < 5km — escala de autódromo).
 */
export function distanceMetersGeo(a, b) {
  if (a == null || b == null) return null;
  if (!Number.isFinite(a.lat) || !Number.isFinite(a.lng)) return null;
  if (!Number.isFinite(b.lat) || !Number.isFinite(b.lng)) return null;
  const dLat = (b.lat - a.lat) * M_PER_DEG_LAT;
  const cosLat = Math.cos((a.lat * Math.PI) / 180);
  const dLng = (b.lng - a.lng) * M_PER_DEG_LAT * cosLat;
  return Math.hypot(dLat, dLng);
}

// ─── Detecção de eventos no stream do trecho ────────────────

/**
 * Retorna o índice da PRIMEIRA amostra em que o piloto começou a frear.
 * Estratégia híbrida (espelha fase-curva.js detectarCortes):
 *  1. Se houver accLong: primeiro com accLong < BRAKE_G_THRESHOLD
 *  2. Senão (fallback): primeira amostra onde dKmh/dt < -KMH_DROP_PER_SEC
 *      por pelo menos 2 amostras consecutivas (anti-spike).
 * Retorna -1 se nada bater.
 */
export function findBrakingIndex(samples) {
  if (!samples || samples.length < MIN_SAMPLES) return -1;
  const temAcc = samples.some(s => Number.isFinite(s.accLong));
  if (temAcc) {
    for (let i = 0; i < samples.length; i++) {
      const a = samples[i].accLong;
      if (Number.isFinite(a) && a < BRAKE_G_THRESHOLD) return i;
    }
    return -1;
  }
  // Fallback sem IMU
  let consecutivos = 0;
  for (let i = 1; i < samples.length; i++) {
    const k0 = samples[i - 1].kmh, k1 = samples[i].kmh;
    const dt = (samples[i].t - samples[i - 1].t) / 1000;
    if (k0 == null || k1 == null || dt <= 0) { consecutivos = 0; continue; }
    const queda = (k0 - k1) / dt;          // km/h por segundo
    if (queda > KMH_DROP_PER_SEC) {
      consecutivos++;
      if (consecutivos >= 2) return i - 1;
    } else {
      consecutivos = 0;
    }
  }
  return -1;
}

/**
 * Retorna o índice da primeira amostra em que o piloto começou a girar
 * (turn-in). Híbrido: gyroAlpha (preferido) ou delta de course.
 */
export function findTurnInIndex(samples) {
  if (!samples || samples.length < MIN_SAMPLES) return -1;
  const temGyro = samples.some(s => Number.isFinite(s.gyroAlpha));
  if (temGyro) {
    for (let i = 0; i < samples.length; i++) {
      const g = samples[i].gyroAlpha;
      if (Number.isFinite(g) && Math.abs(g) > YAW_THRESHOLD_DEG_S) return i;
    }
    return -1;
  }
  // Fallback via course
  for (let i = 1; i < samples.length; i++) {
    const c0 = samples[i - 1].course, c1 = samples[i].course;
    if (c0 == null || c1 == null) continue;
    let d = Math.abs(c1 - c0);
    if (d > 180) d = 360 - d;             // wrap
    if (d > COURSE_DELTA_DEG) return i;
  }
  return -1;
}

/**
 * Retorna o índice da primeira amostra após o apex em que o piloto
 * voltou a acelerar de forma clara.
 *
 * @param {Array} samples
 * @param {number} apexIdx — índice do apex (velocidade mínima); -1 desativa
 */
export function findThrottleIndex(samples, apexIdx) {
  if (!samples || samples.length < MIN_SAMPLES) return -1;
  const start = apexIdx >= 0 ? apexIdx : 0;
  for (let i = start; i < samples.length; i++) {
    const a = samples[i].accLong;
    if (Number.isFinite(a) && a > THROTTLE_G_THRESHOLD) return i;
  }
  return -1;
}

/**
 * Encontra o apex (velocidade mínima) — espelha fase-curva.js.
 */
export function findApexIndex(samples) {
  if (!samples || samples.length < MIN_SAMPLES) return -1;
  let best = -1, vMin = Infinity;
  for (let i = 0; i < samples.length; i++) {
    const k = samples[i].kmh;
    if (Number.isFinite(k) && k < vMin) { vMin = k; best = i; }
  }
  return best;
}

// ─── Avaliação de um trecho ────────────────────────────────

/**
 * @typedef {Object} TrajectoryReport
 * @property {boolean} disponivel
 * @property {string|null} razao              motivo se !disponivel
 * @property {number|null} brakingPointDeviationM
 * @property {number|null} turnInPointDeviationM
 * @property {number|null} throttlePointDeviationM
 * @property {number|null} entryDeviationM
 * @property {number|null} apexDeviationM
 * @property {string} confidence              alta|media|baixa
 * @property {object} debug                   índices encontrados
 */

/**
 * Avalia um trecho contra a referência. Cada métrica é calculada de
 * forma independente — uma falha de detecção (ex: sem accLong na ref)
 * não invalida as outras.
 *
 * @param {object} args
 * @param {Array}  args.samples       amostras do trecho na volta atual
 * @param {Array}  args.refSamples    amostras do MESMO trecho na volta-ref
 * @returns {TrajectoryReport}
 */
export function evaluateSegmentTrajectory({ samples, refSamples } = {}) {
  const empty = (razao) => ({
    disponivel: false, razao,
    brakingPointDeviationM: null, turnInPointDeviationM: null,
    throttlePointDeviationM: null, entryDeviationM: null, apexDeviationM: null,
    confidence: Confidence.BAIXA,
    debug: {},
  });
  if (!samples?.length || samples.length < MIN_SAMPLES) return empty('sem-amostras-atuais');
  if (!refSamples?.length || refSamples.length < MIN_SAMPLES) return empty('sem-amostras-referencia');

  const apexIdx = findApexIndex(samples);
  const refApexIdx = findApexIndex(refSamples);

  const brakingIdx = findBrakingIndex(samples);
  const refBrakingIdx = findBrakingIndex(refSamples);

  const turnInIdx = findTurnInIndex(samples);
  const refTurnInIdx = findTurnInIndex(refSamples);

  const throttleIdx = findThrottleIndex(samples, apexIdx);
  const refThrottleIdx = findThrottleIndex(refSamples, refApexIdx);

  const safeDist = (idxA, arrA, idxB, arrB) => {
    if (idxA < 0 || idxB < 0) return null;
    return distanceMetersGeo(arrA[idxA], arrB[idxB]);
  };

  // Confidence:
  //   ALTA  — todos os 3 eventos críticos detectados em ambos
  //   MEDIA — pelo menos 2 detectados
  //   BAIXA — só apex/entry (geometria pura)
  const detectados = [brakingIdx, refBrakingIdx, turnInIdx, refTurnInIdx,
                      throttleIdx, refThrottleIdx].filter(i => i >= 0).length;
  const confidence = detectados >= 5 ? Confidence.ALTA
                   : detectados >= 3 ? Confidence.MEDIA
                   : Confidence.BAIXA;

  return {
    disponivel: true,
    razao: null,
    brakingPointDeviationM:  safeDist(brakingIdx, samples, refBrakingIdx, refSamples),
    turnInPointDeviationM:   safeDist(turnInIdx, samples, refTurnInIdx, refSamples),
    throttlePointDeviationM: safeDist(throttleIdx, samples, refThrottleIdx, refSamples),
    entryDeviationM:         distanceMetersGeo(samples[0], refSamples[0]),
    apexDeviationM:          safeDist(apexIdx, samples, refApexIdx, refSamples),
    confidence,
    debug: {
      apexIdx, refApexIdx,
      brakingIdx, refBrakingIdx,
      turnInIdx, refTurnInIdx,
      throttleIdx, refThrottleIdx,
    },
  };
}

// ─── Avaliação de successCriteria de uma lição ─────────────

/**
 * Avalia se o piloto cumpriu o successCriteria de L001 (Referência
 * Fixa) — desvio do ponto de freio < toleranciaM em N de M voltas.
 *
 * @param {Array<TrajectoryReport>} reports — uma por volta recente
 * @param {object} cfg
 * @param {number} [cfg.toleranciaM=8]       — desvio aceito
 * @param {number} [cfg.exigidasN=3]
 * @param {number} [cfg.janelaM=4]
 * @returns {{cumpriu: boolean, hits: number, total: number, mediaM: number|null}}
 */
export function evaluateReferenciaFixa(reports, cfg = {}) {
  const toleranciaM = cfg.toleranciaM ?? 8;
  const exigidasN   = cfg.exigidasN ?? 3;
  const janelaM     = cfg.janelaM ?? 4;
  const ultimas = reports.slice(-janelaM);
  let hits = 0, soma = 0, com = 0;
  for (const r of ultimas) {
    const d = r?.brakingPointDeviationM;
    if (d == null) continue;
    com++;
    soma += d;
    if (d < toleranciaM) hits++;
  }
  return {
    cumpriu: hits >= exigidasN,
    hits, total: ultimas.length,
    mediaM: com > 0 ? Math.round((soma / com) * 10) / 10 : null,
  };
}

/**
 * L007 (Curva Cega) — desvio da entrada < toleranciaM nas voltas-ref.
 */
export function evaluateCurvaCega(reports, cfg = {}) {
  const toleranciaM = cfg.toleranciaM ?? 5;
  const exigidasN   = cfg.exigidasN ?? 2;
  const janelaM     = cfg.janelaM ?? 3;
  const ultimas = reports.slice(-janelaM);
  let hits = 0, soma = 0, com = 0;
  for (const r of ultimas) {
    const d = r?.entryDeviationM;
    if (d == null) continue;
    com++;
    soma += d;
    if (d < toleranciaM) hits++;
  }
  return {
    cumpriu: hits >= exigidasN,
    hits, total: ultimas.length,
    mediaM: com > 0 ? Math.round((soma / com) * 10) / 10 : null,
  };
}

// ─── Wrapper de instância (para o caller manter histórico) ─

/**
 * TrajectoryMonitor — instância que mantém histórico por trecho e
 * acumula relatórios. Usar uma instância por sessão de aprendizado.
 */
export class TrajectoryMonitor {
  constructor() {
    /** @type {Map<string, TrajectoryReport[]>} segmentId → reports */
    this.bySegment = new Map();
  }

  /**
   * Registra a finalização de um trecho. O caller deve chamar isso
   * em onSegmentEnd, com os samples acumulados do trecho atual.
   *
   * @returns {TrajectoryReport}
   */
  recordSegment({ segmentId, samples, refSamples } = {}) {
    if (!segmentId) throw new Error('recordSegment: segmentId obrigatório');
    const report = evaluateSegmentTrajectory({ samples, refSamples });
    const arr = this.bySegment.get(segmentId) ?? [];
    arr.push(report);
    this.bySegment.set(segmentId, arr);
    return report;
  }

  /** Histórico de relatórios de um trecho (mais recente por último). */
  getHistory(segmentId) {
    return this.bySegment.get(segmentId)?.slice() ?? [];
  }

  /** Avalia successCriteria da lição contra o histórico do trecho. */
  evaluateLesson({ lessonId, segmentId, cfg = {} } = {}) {
    const reports = this.getHistory(segmentId);
    if (lessonId === 'L001-referencia-fixa') return evaluateReferenciaFixa(reports, cfg);
    if (lessonId === 'L007-curva-cega')      return evaluateCurvaCega(reports, cfg);
    return { cumpriu: false, hits: 0, total: reports.length, mediaM: null,
             razao: 'lesson-sem-avaliador' };
  }

  reset() { this.bySegment.clear(); }
}

File: ios/p1fast-core/Sources/P1FastCore/TrajectoryMonitor.swift
// ═══════════════════════════════════════════════════════════
// TrajectoryMonitor — port Swift de src/domain/trajectory-monitor.js
// ═══════════════════════════════════════════════════════════
// Avaliação de desvio em metros vs volta de referência. Fecha o loop
// de L001 (Referência Fixa) e L007 (Curva Cega) do P1 Coach.
//
// Detecções espelham a lógica JS:
//   • freio   = 1ª amostra com accLong < -0.20g
//               OU (sem IMU) queda sustentada de kmh
//   • giro    = 1ª amostra com |gyroAlpha| > 15°/s
//               OU mudança de course além de 6°
//   • tração  = 1ª amostra após apex com accLong > +0.15g
//
// `Confidence` reusa o enum de Lesson.swift.

import Foundation

public enum TrajectoryMonitorConst {
    public static let brakeGThreshold: Double    = -0.20
    public static let throttleGThreshold: Double = +0.15
    public static let yawThresholdDegS: Double   = 15
    public static let courseDeltaDeg: Double     = 6
    public static let kmhDropPerSec: Double      = 8
    public static let minSamples: Int            = 4
    public static let mPerDegLat: Double         = 111000
}

public enum TrajectoryMonitor_ {

    public struct InputSample: Sendable {
        public let t: Double
        public let lat: Double?
        public let lng: Double?
        public let kmh: Double?
        public let accLong: Double?
        public let gyroAlpha: Double?
        public let course: Double?
        public init(t: Double, lat: Double? = nil, lng: Double? = nil,
                    kmh: Double? = nil, accLong: Double? = nil,
                    gyroAlpha: Double? = nil, course: Double? = nil) {
            self.t = t; self.lat = lat; self.lng = lng
            self.kmh = kmh; self.accLong = accLong
            self.gyroAlpha = gyroAlpha; self.course = course
        }
    }

    public struct Coord: Sendable {
        public let lat: Double?
        public let lng: Double?
    }

    /// Distância plana lat/lng → metros (válido escala autódromo <5km).
    public static func distanceMetersGeo(_ a: InputSample?, _ b: InputSample?) -> Double? {
        guard let a, let b,
              let aLat = a.lat, let aLng = a.lng,
              let bLat = b.lat, let bLng = b.lng,
              aLat.isFinite, aLng.isFinite, bLat.isFinite, bLng.isFinite
        else { return nil }
        let dLat = (bLat - aLat) * TrajectoryMonitorConst.mPerDegLat
        let cosLat = cos(aLat * .pi / 180)
        let dLng = (bLng - aLng) * TrajectoryMonitorConst.mPerDegLat * cosLat
        return (dLat * dLat + dLng * dLng).squareRoot()
    }

    /// Versão para coords arbitrárias (não dentro de InputSample).
    public static func distanceMetersGeo(_ a: (lat: Double, lng: Double)?, _ b: (lat: Double, lng: Double)?) -> Double? {
        guard let a, let b else { return nil }
        let dLat = (b.lat - a.lat) * TrajectoryMonitorConst.mPerDegLat
        let cosLat = cos(a.lat * .pi / 180)
        let dLng = (b.lng - a.lng) * TrajectoryMonitorConst.mPerDegLat * cosLat
        return (dLat * dLat + dLng * dLng).squareRoot()
    }

    public static func findApexIndex(_ samples: [InputSample]) -> Int {
        guard samples.count >= TrajectoryMonitorConst.minSamples else { return -1 }
        var best = -1
        var vMin = Double.infinity
        for (i, s) in samples.enumerated() {
            if let k = s.kmh, k.isFinite, k < vMin {
                vMin = k
                best = i
            }
        }
        return best
    }

    public static func findBrakingIndex(_ samples: [InputSample]) -> Int {
        guard samples.count >= TrajectoryMonitorConst.minSamples else { return -1 }
        let temAcc = samples.contains { $0.accLong != nil && ($0.accLong?.isFinite ?? false) }
        if temAcc {
            for (i, s) in samples.enumerated() {
                if let a = s.accLong, a.isFinite, a < TrajectoryMonitorConst.brakeGThreshold {
                    return i
                }
            }
            return -1
        }
        // Fallback sem IMU
        var consecutivos = 0
        for i in 1..<samples.count {
            guard let k0 = samples[i - 1].kmh, let k1 = samples[i].kmh else {
                consecutivos = 0; continue
            }
            let dt = (samples[i].t - samples[i - 1].t) / 1000.0
            if dt <= 0 { consecutivos = 0; continue }
            let queda = (k0 - k1) / dt
            if queda > TrajectoryMonitorConst.kmhDropPerSec {
                consecutivos += 1
                if consecutivos >= 2 { return i - 1 }
            } else {
                consecutivos = 0
            }
        }
        return -1
    }

    public static func findTurnInIndex(_ samples: [InputSample]) -> Int {
        guard samples.count >= TrajectoryMonitorConst.minSamples else { return -1 }
        let temGyro = samples.contains { $0.gyroAlpha != nil && ($0.gyroAlpha?.isFinite ?? false) }
        if temGyro {
            for (i, s) in samples.enumerated() {
                if let g = s.gyroAlpha, g.isFinite, abs(g) > TrajectoryMonitorConst.yawThresholdDegS {
                    return i
                }
            }
            return -1
        }
        // Fallback via course
        for i in 1..<samples.count {
            guard let c0 = samples[i - 1].course, let c1 = samples[i].course else { continue }
            var d = abs(c1 - c0)
            if d > 180 { d = 360 - d }
            if d > TrajectoryMonitorConst.courseDeltaDeg { return i }
        }
        return -1
    }

    public static func findThrottleIndex(_ samples: [InputSample], apexIdx: Int) -> Int {
        guard samples.count >= TrajectoryMonitorConst.minSamples else { return -1 }
        let start = apexIdx >= 0 ? apexIdx : 0
        for i in start..<samples.count {
            if let a = samples[i].accLong, a.isFinite, a > TrajectoryMonitorConst.throttleGThreshold {
                return i
            }
        }
        return -1
    }

    public struct Report: Sendable {
        public let disponivel: Bool
        public let razao: String?
        public let brakingPointDeviationM: Double?
        public let turnInPointDeviationM: Double?
        public let throttlePointDeviationM: Double?
        public let entryDeviationM: Double?
        public let apexDeviationM: Double?
        public let confidence: Confidence
        public let debug: Debug

        public init(disponivel: Bool, razao: String?,
                    brakingPointDeviationM: Double?, turnInPointDeviationM: Double?,
                    throttlePointDeviationM: Double?, entryDeviationM: Double?,
                    apexDeviationM: Double?,
                    confidence: Confidence, debug: Debug) {
            self.disponivel = disponivel
            self.razao = razao
            self.brakingPointDeviationM = brakingPointDeviationM
            self.turnInPointDeviationM = turnInPointDeviationM
            self.throttlePointDeviationM = throttlePointDeviationM
            self.entryDeviationM = entryDeviationM
            self.apexDeviationM = apexDeviationM
            self.confidence = confidence
            self.debug = debug
        }

        public struct Debug: Sendable {
            public let apexIdx: Int
            public let refApexIdx: Int
            public let brakingIdx: Int
            public let refBrakingIdx: Int
            public let turnInIdx: Int
            public let refTurnInIdx: Int
            public let throttleIdx: Int
            public let refThrottleIdx: Int
            public init(apexIdx: Int, refApexIdx: Int,
                        brakingIdx: Int, refBrakingIdx: Int,
                        turnInIdx: Int, refTurnInIdx: Int,
                        throttleIdx: Int, refThrottleIdx: Int) {
                self.apexIdx = apexIdx
                self.refApexIdx = refApexIdx
                self.brakingIdx = brakingIdx
                self.refBrakingIdx = refBrakingIdx
                self.turnInIdx = turnInIdx
                self.refTurnInIdx = refTurnInIdx
                self.throttleIdx = throttleIdx
                self.refThrottleIdx = refThrottleIdx
            }
        }

        static func empty(_ razao: String) -> Report {
            Report(
                disponivel: false, razao: razao,
                brakingPointDeviationM: nil, turnInPointDeviationM: nil,
                throttlePointDeviationM: nil, entryDeviationM: nil, apexDeviationM: nil,
                confidence: .baixa,
                debug: Debug(apexIdx: -1, refApexIdx: -1,
                             brakingIdx: -1, refBrakingIdx: -1,
                             turnInIdx: -1, refTurnInIdx: -1,
                             throttleIdx: -1, refThrottleIdx: -1)
            )
        }
    }

    public static func evaluateSegmentTrajectory(samples: [InputSample], refSamples: [InputSample]) -> Report {
        guard samples.count >= TrajectoryMonitorConst.minSamples else {
            return Report.empty("sem-amostras-atuais")
        }
        guard refSamples.count >= TrajectoryMonitorConst.minSamples else {
            return Report.empty("sem-amostras-referencia")
        }

        let apexIdx       = findApexIndex(samples)
        let refApexIdx    = findApexIndex(refSamples)
        let brakingIdx    = findBrakingIndex(samples)
        let refBrakingIdx = findBrakingIndex(refSamples)
        let turnInIdx     = findTurnInIndex(samples)
        let refTurnInIdx  = findTurnInIndex(refSamples)
        let throttleIdx   = findThrottleIndex(samples, apexIdx: apexIdx)
        let refThrottleIdx = findThrottleIndex(refSamples, apexIdx: refApexIdx)

        func safeDist(_ idxA: Int, _ arrA: [InputSample], _ idxB: Int, _ arrB: [InputSample]) -> Double? {
            if idxA < 0 || idxB < 0 { return nil }
            return distanceMetersGeo(arrA[idxA], arrB[idxB])
        }

        let detectados = [brakingIdx, refBrakingIdx, turnInIdx, refTurnInIdx,
                          throttleIdx, refThrottleIdx].filter { $0 >= 0 }.count
        let confidence: Confidence = detectados >= 5 ? .alta
                                   : detectados >= 3 ? .media
                                   : .baixa

        return Report(
            disponivel: true, razao: nil,
            brakingPointDeviationM:  safeDist(brakingIdx, samples, refBrakingIdx, refSamples),
            turnInPointDeviationM:   safeDist(turnInIdx, samples, refTurnInIdx, refSamples),
            throttlePointDeviationM: safeDist(throttleIdx, samples, refThrottleIdx, refSamples),
            entryDeviationM:         distanceMetersGeo(samples.first, refSamples.first),
            apexDeviationM:          safeDist(apexIdx, samples, refApexIdx, refSamples),
            confidence: confidence,
            debug: Report.Debug(
                apexIdx: apexIdx, refApexIdx: refApexIdx,
                brakingIdx: brakingIdx, refBrakingIdx: refBrakingIdx,
                turnInIdx: turnInIdx, refTurnInIdx: refTurnInIdx,
                throttleIdx: throttleIdx, refThrottleIdx: refThrottleIdx
            )
        )
    }

    public struct LessonOutcome: Sendable {
        public let cumpriu: Bool
        public let hits: Int
        public let total: Int
        public let mediaM: Double?
        public let razao: String?
    }

    public static func evaluateReferenciaFixa(_ reports: [Report],
                                              toleranciaM: Double = 8,
                                              exigidasN: Int = 3,
                                              janelaM: Int = 4) -> LessonOutcome {
        let ultimas = Array(reports.suffix(janelaM))
        var hits = 0, com = 0
        var soma: Double = 0
        for r in ultimas {
            guard let d = r.brakingPointDeviationM else { continue }
            com += 1
            soma += d
            if d < toleranciaM { hits += 1 }
        }
        let media: Double? = com > 0 ? (soma / Double(com) * 10).rounded() / 10 : nil
        return LessonOutcome(cumpriu: hits >= exigidasN, hits: hits,
                             total: ultimas.count, mediaM: media, razao: nil)
    }

    public static func evaluateCurvaCega(_ reports: [Report],
                                         toleranciaM: Double = 5,
                                         exigidasN: Int = 2,
                                         janelaM: Int = 3) -> LessonOutcome {
        let ultimas = Array(reports.suffix(janelaM))
        var hits = 0, com = 0
        var soma: Double = 0
        for r in ultimas {
            guard let d = r.entryDeviationM else { continue }
            com += 1
            soma += d
            if d < toleranciaM { hits += 1 }
        }
        let media: Double? = com > 0 ? (soma / Double(com) * 10).rounded() / 10 : nil
        return LessonOutcome(cumpriu: hits >= exigidasN, hits: hits,
                             total: ultimas.count, mediaM: media, razao: nil)
    }
}

/// Wrapper de instância que mantém histórico por trecho.
public final class TrajectoryMonitor {
    private var bySegment: [String: [TrajectoryMonitor_.Report]] = [:]
    public init() {}

    @discardableResult
    public func recordSegment(segmentId: String,
                              samples: [TrajectoryMonitor_.InputSample],
                              refSamples: [TrajectoryMonitor_.InputSample]) -> TrajectoryMonitor_.Report {
        precondition(!segmentId.isEmpty, "recordSegment: segmentId obrigatório")
        let report = TrajectoryMonitor_.evaluateSegmentTrajectory(samples: samples, refSamples: refSamples)
        var arr = bySegment[segmentId] ?? []
        arr.append(report)
        bySegment[segmentId] = arr
        return report
    }

    public func getHistory(_ segmentId: String) -> [TrajectoryMonitor_.Report] {
        return bySegment[segmentId] ?? []
    }

    public func evaluateLesson(lessonId: String, segmentId: String) -> TrajectoryMonitor_.LessonOutcome {
        let reports = getHistory(segmentId)
        switch lessonId {
        case "L001-referencia-fixa":
            return TrajectoryMonitor_.evaluateReferenciaFixa(reports)
        case "L007-curva-cega":
            return TrajectoryMonitor_.evaluateCurvaCega(reports)
        default:
            return TrajectoryMonitor_.LessonOutcome(
                cumpriu: false, hits: 0, total: reports.count,
                mediaM: nil, razao: "lesson-sem-avaliador"
            )
        }
    }

    public func reset() { bySegment.removeAll() }
}
